File: LeetCode/ValidBinarySearchTree.java


/**
 * Created by yi on 4/24/15.
 */
public class ValidBinarySearchTree {
    public boolean isValidBST(TreeNode root) {
        return isValidNode(root, null, null);
    }

    private boolean isValidNode(TreeNode root, Integer max, Integer min) {
        if (root == null) return true;
        return (min == null || root.val > min) && (max == null || root.val < max) && isValidNode(root.left, root.val, min) && isValidNode(root.right, max, root.val);
    }

    private class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode(int x) {
            this.val = x;
        }
    }
}

File: LeetCode/Python/ValidBinarySearchTree.py
__author__ = 'yi'


class TreeNode:
    def __init__(self,x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def validNode(self, root, maxval, minval):
        if root is None:
            return True
        return (maxval is None or root.val < maxval) and (minval is None or root.val > minval)\
            and self.validNode(root.left, root.val, minval) and self.validNode(root.right, maxval, root.val)

    def isValidBST(self, root):
        return self.validNode(root, None, None)
